Sources/ShapedArray/Basic.swift

```diff
@@ -49,6 +53,115 @@ extension ShapedArray {
         self.reshaped(to: -1)
     }
 
+    // helper data structure for subarray generation
+    @usableFromInline
+    internal struct SubarrayIndices {
+        @usableFromInline
+        let start: [Int]
+        @usableFromInline
+        let end: [Int]
+    }
+
+    // helper function for subarray indices generation
+    /// takes a shape for the original array, a shape for the subarray, 
+    /// and an axis on which the original array would be divided 
+    /// and returns an array of start and end indices for each subarray
+
+    @usableFromInline
+    internal func _calculateSubarrayIndices(shape: [Int], subarrayShape: [Int], axis: Int) -> [SubarrayIndices] {
+        let numberOfSubarrays = shape[axis] / subarrayShape[axis]
+        
+        var subarrays = [SubarrayIndices]()
+        
+        for i in 0..<numberOfSubarrays {
+            var startIndices = [Int](repeating: 0, count: shape.count)
+            var endIndices = shape
+            
+            startIndices[axis] = i * subarrayShape[axis]
+            endIndices[axis] = startIndices[axis] + subarrayShape[axis]
+            
+            subarrays.append(SubarrayIndices(start: startIndices, end: endIndices))
+        }
+        
+        return subarrays
+    }
+
+    // helper function for subarray indices generation
+    /// takes a start and end index for each dimension and returns all the n-dim indices in between
+    @usableFromInline
+    internal func _generateIndices(start: [Int], end: [Int], currentIndex: [Int] = [], depth: Int = 0) -> [[Int]] {
+        if depth == start.count {
+            return [currentIndex]
+        }
+        
+        var indices = [[Int]]()
+        
+        for i in start[depth]..<end[depth] {
+            let newIndex = currentIndex + [i]
+            let subIndices = _generateIndices(start: start, end: end, currentIndex: newIndex, depth: depth + 1)
```

Comment:
This isn't tail recursive so there's no tail-call optimization. Granted probably plenty to optimize in the future, but maybe add a comment here calling that out explicitly?
E.g. `TODO: Possible tail-call optimization`?